Aniyomi crashes when media button from headset is clicked

### Steps to reproduce

1. Connect a bluetooth earphone to device
2. Open aniyomi
3. Press play / pause button

### Expected behavior

Priority wise
1. App should not crash
2. Should play / pause the video if currently playing
3. <TBD> Should ignore the event if no video is in front

### Actual behavior

App crashes

### Crash logs

07-31 10:50:13.069  2585  2585 E AndroidRuntime: FATAL EXCEPTION: main
07-31 10:50:13.069  2585  2585 E AndroidRuntime: Process: xyz.jmir.tachiyomi.mi, PID: 2585
07-31 10:50:13.069  2585  2585 E AndroidRuntime: java.lang.RuntimeException: Unable to start receiver androidx.media.session.MediaButtonReceiver: java.lang.IllegalStateException: Could not find any Service that handles android.intent.action.MEDIA_BUTTON or implements a media browser service.
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3047)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.app.ActivityThread.-wrap18(ActivityThread.java)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1561)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6119)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: Caused by: java.lang.IllegalStateException: Could not find any Service that handles android.intent.action.MEDIA_BUTTON or implements a media browser service.
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at androidx.media.session.MediaButtonReceiver.onReceive(MediaButtonReceiver.java:15)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3040)
07-31 10:50:13.069  2585  2585 E AndroidRuntime: 	... 8 more

### Aniyomi version

0.12.3.10

### Android version

Android 7

### Device

Redmi Note 5

### Other details

Did Google search and this issue seems to have workaround.

https://stackoverflow.com/questions/51216279/app-crash-on-headset-button-click

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with an extension, I should be opening an issue in the [extensions repository](https://github.com/aniyomiorg/aniyomi-extensions/issues/new/choose).
- [X] I have tried the [troubleshooting guide](https://aniyomi.org/help/guides/troubleshooting/).
- [X] I have updated the app to version **[0.12.3.10](https://github.com/aniyomiorg/aniyomi/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I will fill out all of the requested information in this form.

--- a/app/build.gradle.kts
+++ b/app/build.gradle.kts
@@ -187,6 +187,7 @@ dependencies {
     implementation(androidx.recyclerview)
     implementation(androidx.viewpager)
     implementation(androidx.profileinstaller)
+    implementation(androidx.mediasession)
 
     implementation(androidx.bundles.lifecycle)
 

--- a/app/src/main/java/eu/kanade/tachiyomi/ui/player/PlayerActivity.kt
+++ b/app/src/main/java/eu/kanade/tachiyomi/ui/player/PlayerActivity.kt
@@ -16,6 +16,9 @@ import android.os.Bundle
 import android.os.Handler
 import android.os.Looper
 import android.os.ParcelFileDescriptor
+import android.support.v4.media.session.MediaControllerCompat
+import android.support.v4.media.session.MediaSessionCompat
+import android.support.v4.media.session.PlaybackStateCompat
 import android.util.DisplayMetrics
 import android.view.KeyEvent
 import android.view.MotionEvent
@@ -145,6 +148,10 @@ class PlayerActivity : BaseActivity() {
     private var mReceiver: BroadcastReceiver? = null
 
     lateinit var binding: PlayerActivityBinding
+
+    private lateinit var mediaSession: MediaSessionCompat
+
+    private val playbackStateBuilder = PlaybackStateCompat.Builder()
 
     internal val player get() = binding.player
 
@@ -276,6 +283,7 @@ class PlayerActivity : BaseActivity() {
         super.onCreate(savedInstanceState)
 
         setupPlayerControls()
+        setupMediaSession()
         setupPlayerMPV()
         setupPlayerAudio()
         setupPlayerBrightness()
@@ -434,6 +442,89 @@ class PlayerActivity : BaseActivity() {
     }
 
     @Suppress("DEPRECATION")
+    private fun setupMediaSession() {
+        mediaSession = MediaSessionCompat(this, "Aniyomi_Player_Session").apply {
+            // Enable callbacks from MediaButtons and TransportControls
+            setFlags(
+                MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or
+                    MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS,
+            )
+
+            // Do not let MediaButtons restart the player when the app is not visible
+            setMediaButtonReceiver(null)
+
+            setPlaybackState(
+                with(playbackStateBuilder) {
+                    setState(
+                        PlaybackStateCompat.STATE_NONE,
+                        0L,
+                        0.0F,
+                    )
+                    build()
+                },
+            )
+
+            // Implement methods that handle callbacks from a media controller
+            setCallback(
+                object : MediaSessionCompat.Callback() {
+                    override fun onPlay() {
+                        player.paused = false
+                        playerControls.toggleControls(isTapped = true)
+                        updatePlaybackState()
+                    }
+
+                    override fun onPause() {
+                        player.paused = true
+                        playerControls.toggleControls()
+                        updatePlaybackState(pause = true)
+                    }
+
+                    override fun onSkipToPrevious() {
+                        changeEpisode(viewModel.getAdjacentEpisodeId(previous = true))
+                    }
+
+                    override fun onSkipToNext() {
+                        changeEpisode(viewModel.getAdjacentEpisodeId(previous = false))
+                    }
+                },
+            )
+        }
+
+        MediaControllerCompat(this, mediaSession).also { mediaController ->
+            MediaControllerCompat.setMediaController(this, mediaController)
+        }
+    }
+
+    private fun updatePlaybackState(cachePause: Boolean = false, pause: Boolean = false) {
+        val state = when {
+            player.timePos?.let { it < 0 } ?: true ||
+                player.duration?.let { it <= 0 } ?: true -> PlaybackStateCompat.STATE_CONNECTING
+            cachePause -> PlaybackStateCompat.STATE_BUFFERING
+            pause or (player.paused == true) -> PlaybackStateCompat.STATE_PAUSED
+            else -> PlaybackStateCompat.STATE_PLAYING
+        }
+        var actions = PlaybackStateCompat.ACTION_PLAY or
+            PlaybackStateCompat.ACTION_PLAY_PAUSE or
+            PlaybackStateCompat.ACTION_PAUSE
+        if (viewModel.currentPlaylist.size > 1) {
+            actions = actions or PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or
+                PlaybackStateCompat.ACTION_SKIP_TO_NEXT
+        }
+
+        mediaSession.setPlaybackState(
+            with(playbackStateBuilder) {
+                setState(
+                    state,
+                    player.timePos?.toLong() ?: 0L,
+                    player.playbackSpeed?.toFloat() ?: 1.0f,
+                )
+                setActions(actions)
+                build()
+            },
+        )
+    }
+
+    @Suppress("DEPRECATION")
     private fun loadDeviceDimensions() {
         this@PlayerActivity.requestedOrientation = playerPreferences.defaultPlayerOrientationType().get()
         val dm = DisplayMetrics()
@@ -497,6 +588,9 @@ class PlayerActivity : BaseActivity() {
     }
 
     override fun onDestroy() {
+        mediaSession.isActive = false
+        mediaSession.release()
+
         playerPreferences.playerVolumeValue().set(fineVolume)
         playerPreferences.playerBrightnessValue().set(brightness)
         MPVLib.removeLogObserver(playerObserver)
@@ -1560,19 +1654,28 @@ class PlayerActivity : BaseActivity() {
             "time-pos" -> {
                 playerControls.updatePlaybackPos(value.toInt())
                 viewModel.viewModelScope.launchUI { aniSkipStuff(value) }
+                updatePlaybackState()
             }
-            "duration" -> playerControls.updatePlaybackDuration(value.toInt())
+            "duration" -> {
+                playerControls.updatePlaybackDuration(value.toInt())
+                mediaSession.isActive = true
+                updatePlaybackState()
+            }
         }
     }
 
     internal fun eventPropertyUi(property: String, value: Boolean) {
         when (property) {
             "seeking" -> isSeeking(value)
-            "paused-for-cache" -> showLoadingIndicator(value)
+            "paused-for-cache" -> {
+                showLoadingIndicator(value)
+                updatePlaybackState(cachePause = true)
+            }
             "pause" -> {
                 if (!isFinishing) {
                     setAudioFocus(value)
                     updatePlaybackStatus(value)
+                    updatePlaybackState(pause = true)
                 }
             }
             "eof-reached" -> endFile(value)

--- a/gradle/androidx.versions.toml
+++ b/gradle/androidx.versions.toml
@@ -16,6 +16,7 @@ recyclerview = "androidx.recyclerview:recyclerview:1.3.0"
 viewpager = "androidx.viewpager:viewpager:1.1.0-alpha01"
 glance = "androidx.glance:glance-appwidget:1.0.0-alpha03"
 profileinstaller = "androidx.profileinstaller:profileinstaller:1.3.0"
+mediasession = "androidx.media:media:1.6.0"
 
 lifecycle-common = { module = "androidx.lifecycle:lifecycle-common", version.ref = "lifecycle_version" }
 lifecycle-process = { module = "androidx.lifecycle:lifecycle-process", version.ref = "lifecycle_version" }

